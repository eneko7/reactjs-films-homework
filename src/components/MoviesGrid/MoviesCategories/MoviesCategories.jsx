import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import queryString from 'query-string';
import style from './MoviesCategories.scss';
import GenresContainer from '../Genres/GenresContainer';

class MoviesCategories extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      activeCategory: 'Trending',
    };
    this.fetchFilmsByCategory = this.fetchFilmsByCategory.bind(this);
  }

  componentDidMount() {
    const { location: { search, pathname } } = this.props;
    const parsed = queryString.parse(search);
    const { sort } = parsed;
    if (pathname === '/') {
      this.fetchFilmsByCategory('Trending');
    }
    if (pathname === '/films') {
      this.fetchFilmsByCategory(sort);
    }
    if (pathname === '/genres') {
      this.setState(() => ({
        activeCategory: 'Genres',
      }));
    }
  }

  fetchFilmsByCategory(el) {
    this.setState(() => ({
      activeCategory: el,
    }));
    if (el !== 'Genres') {
      const { pushNavigationLink } = this.props;
      pushNavigationLink(el);
    }
  }

  render() {
    const categories = [
      {
        route: '/films?sort=Trending',
        query: 'trending',
        name: 'Trending',
        defaultActive: true,
      },
      {
        route: '/films?sort=Top Rated',
        query: 'top-rated',
        name: 'Top Rated',
        defaultActive: false,
      },
      {
        route: '/films?sort=Coming Soon',
        query: 'coming-soon',
        name: 'Coming Soon',
        defaultActive: false,
      },
    ];
    const { activeCategory } = this.state;
    const active = `${style.active}`;
    const categoritesBlock = categories.map((el) => {
      if (el.name !== activeCategory) {
        return (
          <div key={el.name} className={style.moviesGrid_categories_item}>
            <Link to={el.route}>
              <button type="button" data-filter={el.name} className={`${style.moviesGrid_categories_item_button}`} onClick={this.fetchFilmsByCategory.bind(null, el.name)}>{el.name}</button>
            </Link>
          </div>
        );
      }
      return (
        <div key={el.name} className={style.moviesGrid_categories_item}>
          <Link to={el.route}>
            <button type="button" data-filter={el.name} className={`${active} ${style.moviesGrid_categories_item_button}`} onClick={this.fetchFilmsByCategory.bind(null, el.name)}>{el.name}</button>
          </Link>
        </div>
      );
    });
    return (
      <div className={style.moviesGrid_categories}>
        {categoritesBlock}
        <div className={style.moviesGrid_categories_item}>
          <GenresContainer title="Genres" activeCategory={activeCategory} fetchChange={this.fetchFilmsByCategory} />
        </div>
      </div>
    );
  }
}
MoviesCategories.propTypes = {
  pushNavigationLink: PropTypes.func.isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string,
    search: PropTypes.string,
  }).isRequired,
};
export default MoviesCategories;
