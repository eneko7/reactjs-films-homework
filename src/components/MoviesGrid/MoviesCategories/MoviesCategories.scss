.moviesGrid_categories {
  margin: 70px 7%;
  margin-top: 0;
  margin-bottom: 30px;
  display: flex;
  justify-content: flex-start;

  .moviesGrid_categories_item {
    .moviesGrid_categories_item_button {
      cursor: pointer;
      font-size: 25px;
      line-height: 30px;
      color: #0aaee4;
      font-family: Helvetica, Arial, Verdana, Tahoma, sans-serif;
      margin-right: 30px;
      padding-bottom: 5px;
      text-align: center;
      background: none;
      border:none;
    }
    .moviesGrid_categories_item_button:focus {
      outline: none;
    }
    .moviesGrid_categories_item_button.active{
      border-bottom:4px solid #0aaee4;
    }
  }
}