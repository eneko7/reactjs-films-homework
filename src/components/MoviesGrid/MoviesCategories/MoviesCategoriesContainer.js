import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import MoviesCategories from './MoviesCategories';
import { pushNavigationLink } from '../../../modules/navlinks/navlinksActions';
import getClickedLink from '../../../modules/navlinks/navlinksSelectors';

const mapStateToProps = state => ({
  clickedLink: getClickedLink(state),
});

const mapDispatchToProps = {
  pushNavigationLink,
};
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(MoviesCategories));
