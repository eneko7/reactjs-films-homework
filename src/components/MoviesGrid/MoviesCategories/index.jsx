import MoviesCategories from './MoviesCategoriesContainer';

export default MoviesCategories;
