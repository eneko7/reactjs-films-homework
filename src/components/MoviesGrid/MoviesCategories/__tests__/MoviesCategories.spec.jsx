import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import TestRenderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import MoviesCategories from '../index';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore({
  navlinks: {
    clickedLink: 'Trending',
  },
  genres: {
    allGenres: [
      {
        id: 1,
        name: 'adventure',
      },
      {
        id: 2,
        name: 'action',
      },
      {
        id: 3,
        name: 'drama',
      },
    ],
    lastGenreID: 1,
  },
});
jest.mock('../../Genres/GenresContainer', () => () => <div>test</div>);
const shallow = new ShallowRenderer();
describe('MoviesCategories Snapshot', () => {
  it('renders', () => {
    const component = shallow.render(
      <Provider store={store}>
        <MoviesCategories />
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });

  it('renders categories', () => {
    const component = TestRenderer.create(
      <Provider store={store}>
        <MoviesCategories />
      </Provider>,
    );
    const elArr = ['Trending', 'Top Rated', 'Coming Soon'];
    elArr.map(el => component.root.findByProps({ 'data-filter': el }).props.onClick(el));
    expect(component).toMatchSnapshot();
  });

  it('renders genres', () => {
    const component = TestRenderer.create(
      <Provider store={store}>
        <MoviesCategories />
      </Provider>,
    );
    const { instance } = component.root;
    const el = 'Genres';
    instance.setState(() => ({
      activeCategory: el,
    }));
    component.root.findByProps({ 'data-filter': el }).props.onClick(el);
    expect(component).toMatchSnapshot();
  });
});
