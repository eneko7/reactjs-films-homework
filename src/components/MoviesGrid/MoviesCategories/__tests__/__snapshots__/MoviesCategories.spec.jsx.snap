// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesCategories Snapshot renders 1`] = `<withRouter(Connect(MoviesCategories)) />`;

exports[`MoviesCategories Snapshot renders categories 1`] = `
Array [
  <div
    className="moviesGrid_categories_item"
  >
    <button
      className="moviesGrid_categories_item_button"
      data-filter="Trending"
      onClick={[Function]}
      type="button"
    >
      Trending
    </button>
  </div>,
  <div
    className="moviesGrid_categories_item"
  >
    <button
      className="moviesGrid_categories_item_button"
      data-filter="Top Rated"
      onClick={[Function]}
      type="button"
    >
      Top Rated
    </button>
  </div>,
  <div
    className="moviesGrid_categories_item"
  >
    <button
      className="active moviesGrid_categories_item_button"
      data-filter="Coming Soon"
      onClick={[Function]}
      type="button"
    >
      Coming Soon
    </button>
  </div>,
  <div
    className="moviesGrid_categories_item"
  >
    <div>
      test
    </div>
  </div>,
]
`;

exports[`MoviesCategories Snapshot renders genres 1`] = `
Array [
  <div
    className="moviesGrid_categories_item"
  >
    <button
      className="active moviesGrid_categories_item_button"
      data-filter="Trending"
      onClick={[Function]}
      type="button"
    >
      Trending
    </button>
  </div>,
  <div
    className="moviesGrid_categories_item"
  >
    <button
      className="moviesGrid_categories_item_button"
      data-filter="Top Rated"
      onClick={[Function]}
      type="button"
    >
      Top Rated
    </button>
  </div>,
  <div
    className="moviesGrid_categories_item"
  >
    <button
      className="moviesGrid_categories_item_button"
      data-filter="Coming Soon"
      onClick={[Function]}
      type="button"
    >
      Coming Soon
    </button>
  </div>,
  <div
    className="moviesGrid_categories_item"
  >
    <div>
      test
    </div>
  </div>,
]
`;
